### PR TITLE
Fix in readme and CMake build system

### LINKED ISSUE #7649: How do I add CBMC as a dependency in a cmake project using FetchContent?
I am currently implementing something on top of a stub created by Elizabeth Polgreen based on Makefiles. As I am mostly used to Clion (and its IDE features) I was trying to use CMAKE to handle all dependencies etc. However, when using cmake's [fetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html) I ran into a couple of issues.
The simplest one was the fact that when compiling the "-W-return-type" flag failed, which I fixed by adding 
`set (CMAKE_CXX_FLAGS "-Wno-return-type")`

However, the issues I am not able to fix are the ones surrounding the scripts such as `bash-autocomplete/extract_switches.sh`.

So I was wondering if anyone else is using CMAKE/fetch_content and has anything to add. Also, here is the CMakeFile I am using to reproduce the errors:

```
cmake_minimum_required(VERSION 3.24)
project(CBMCTest)
include(FetchContent)

set(CMAKE_CXX_STANDARD 20)

set (CMAKE_CXX_FLAGS "-Wno-return-type")

FetchContent_Declare(cbmc
        GIT_REPOSITORY https://github.com/diffblue/cbmc
        GIT_TAG develop
        )
FetchContent_MakeAvailable(cbmc)

add_executable(CBMCTest main.cpp)

#probably some target_link_library(CBMCTest cprover goto .... etc.)

```

CBMC version: devel
Operating system: Linux kernel 6.2.8-1-MANJARO / Manjaro

### PR DESCRIPTION
**Pull request by jparsert**
Fixes #7649 for Cmake builds using `add_subdirectory` as well as `fetch_content`.
Now the bash scripts are generated/called relative to the directory where `/src/cbmc/CMakeFiles.txt` is located. This means that when the original `CMakeFiles.txt` is outside of cbmc's root directory (for example in an external project) then the correct path is used when running the bash scripts.

Edit( This has been removed and a separate issue has been created): I also added the "-Wno-return-type" flag as that threw compilation errors in cases where `UNREACHABLE` was used but no return was used in dead code.

I also fixed a link in the readme.md that used to point to "MINI-PROJECTS". It now points to "FEATURE_IDEAS.md".

### CODE REVIEW

File: CMakeLists.txt
Comment (NlightNFotis) [inline CMakeLists.txt]: I'm not sure if this change is _correct_ - the commit message indicates that this would trigger on code marked with our internal `UNREACHABLE` macro (unless I've misunderstood it), but these are _supposed_ to behave in the way they do to indicate a path that the developer intended to be unreachable (and fail loudly if they are not).

File: README.md
Comment (NlightNFotis) [inline README.md:110]: 👍🏻

File: src/cbmc/CMakeLists.txt
Comment (NlightNFotis) [inline src/cbmc/CMakeLists.txt]: I want to defer checking the correctness of that to @esteffin

**Review by NlightNFotis — COMMENTED** [review]
Some questions.

As an aside, could you please remove the last commit?

The way we work in CBMC is we `rebase` our branches on top of branch `develop`, rather than using merge commits (outside of PRs getting merged).

That should be an easy fix - `rebase -i HEAD~4` and `drop` the last commit, then `git checkout develop && git pull origin develop` followed by a `git checkout <yourbranch> && git rebase develop`. (**EDIT**: A faster way might be `git pull origin develop --rebase` while on top of your branch, but I'm not sure if this equivalent. A slightly more adventurous option if you're up for it).

File: CMakeLists.txt
Comment (jparsert) [reply CMakeLists.txt]: The problem I get when compiling is sometimes there the function ends in a `UNREACHABLE` macro but not a return statement. Since GCC does not know that this is unreachable, it complains that there is a path in the non-void function that does not end in a `return` statement. Hence, it throws a compiler warning, an since `-Wall` is set, this warning is actually an error and compilation fails.     
An other way to resolve this is by forcing to also add a return statement even when the `unreachable` makro is used.

Comment (jparsert) [reply CMakeLists.txt]: I removed the last commit. Hope everything works out now.

Comment (thomasspriggs) [reply CMakeLists.txt]: The problem you explain with `UNREACHABLE` and return warnings isn't one I have encountered when building with GCC myself. The control flow is such that encountering an `UNREACHABLE` will either result in an exception being thrown or the termination of the process. Either way execution does not continue after the `UNREACHABLE` which means that neither return statements or suppressing the warnings should be required. Can you give us any further information about the platform and build options which are required to reproduce this issue?

**Comment by thomasspriggs** [comment]
A quick general comment on this PR is that it would be easier to manage on our side if it was 2 PRs. The clue to this is that you needed the word "and" in the title. The documentation fix is fairly straight forwardly an improvement, which could be merged straight away. But the changes to the build system require slightly more scrutiny, potentially from different reviewers with different expertise. It isn't worth going to the effort of splitting this PR into 2 PRs right now in my opinion. But it would be good to consider next time though.

File: CMakeLists.txt
Comment (jparsert) [reply CMakeLists.txt]: When building CBMC in my IDE (clion) I get the following error:

```
/usr/bin/c++ -DHAVE_MINISAT2 -DSATCHECK_MINISAT2 -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/cbmc/cmake-build-debug/src -I/home/cbmc/src -I/home/cbmc/cmake-build-debug/src/solvers -I/home/cbmc/src/solvers -I/home/cbmc/cmake-build-debug/minisat2-src -I/home/cbmc/cmake-build-debug/src/util -I/home/cbmc/src/util -I/home/cbmc/cmake-build-debug/src/big-int -I/home/cbmc/src/big-int -I/home/cbmc/cmake-build-debug/src/langapi -I/home/cbmc/src/langapi -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wswitch-enum -g -fdiagnostics-color=always -std=gnu++11 -MD -MT src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -MF src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o.d -o src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -c /home/cbmc/src/solvers/flattening/boolbv_get.cpp
/home/cbmc/src/solvers/flattening/boolbv_get.cpp: In member function ‘virtual exprt boolbvt::bv_get_rec(const exprt&, const bvt&, std::size_t) const’:
/home/cbmc/src/solvers/flattening/boolbv_get.cpp:272:1: error: control reaches end of non-void function [-Werror=return-type]
  272 | }
      | ^
cc1plus: all warnings being treated as errors
```

EDIT 2:
See below on how to reproduce this issue

Comment (jparsert) [reply CMakeLists.txt]: Alright, here is how I reproduce the issues with gcc 12.2.1:

```
git clone git@github.com:diffblue/cbmc.git
cd cbmc
git submodule update --init
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_MAKE_PROGRAM=/path/to/ninja -G Ninja -S /home/cbmc -B /home/cbmc/build
cmake --build . --target all -j 6

```

If you are unable to reproduce it I can create a docker container

Comment (thomasspriggs) [reply CMakeLists.txt]: That is helpful. Can you also confirm which OS flavour / version you have observed this issue with?

Comment (jparsert) [reply CMakeLists.txt]: Manjaro Linux. Kernel 6.2.9, worth noting that I had that issue some time ago as well but back then I did not end up using cbmc for my project so I never went looking. Also, I just looked and it seems to not be an issue with ninja, as I also get it when using make instead of ninja.

Comment (NlightNFotis) [reply CMakeLists.txt]: @jparsert I see on your compiler invocation that it's building with `-std=gnu++11`.

I'm not sure if this is at fault here - could you try forcing it to build with `-std=c++11` and see what happens?

(Sorry, blueskying here - can't be more helpful than that as I don't use GCC or Linux myself).

Comment (jparsert) [reply CMakeLists.txt]: That's a good suggestion, but changing it did not solve the error. I get the same error with the following compilation command: 
`/usr/bin/c++ -DHAVE_MINISAT2 -DSATCHECK_MINISAT2 -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/julian/Documents/cbmc/cmake-build-debug/src -I/home/julian/Documents/cbmc/src -I/home/julian/Documents/cbmc/cmake-build-debug/src/solvers -I/home/julian/Documents/cbmc/src/solvers -I/home/julian/Documents/cbmc/cmake-build-debug/minisat2-src -I/home/julian/Documents/cbmc/cmake-build-debug/src/util -I/home/julian/Documents/cbmc/src/util -I/home/julian/Documents/cbmc/cmake-build-debug/src/big-int -I/home/julian/Documents/cbmc/src/big-int -I/home/julian/Documents/cbmc/cmake-build-debug/src/langapi -I/home/julian/Documents/cbmc/src/langapi -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wswitch-enum -g -fdiagnostics-color=always -std=c++11 -MD -MT src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -MF src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o.d -o src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -c /home/julian/Documents/cbmc/src/solvers/flattening/boolbv_get.cpp
`

File: src/cbmc/CMakeLists.txt
Comment (esteffin) [reply src/cbmc/CMakeLists.txt]: This solves the problem, however I would suggest a tidier approach that avoids the `../..`.

Why not adding a variable `CBMC_ROOT_DIR` to the `CMakeLists.txt` with the value of `${CMAKE_CURRENT_SOURCE_DIR}`, and then refer to that from `src/cbmc/CMakeLists.txt`?

This will remove all the `..`s and also will be usable from other folders in case it is needed.

File: CMakeLists.txt
Comment (esteffin) [reply CMakeLists.txt]: I managed to reproduce the issue here.

GCC 12 fails to build CBMC when using `Debug` as CMake build type.

I created an issue about this [here](https://github.com/diffblue/cbmc/issues/7663).

I suggest to change the build configuration to something different than `Debug` on CLion while the issue is being addressed.

Also I would suggest removing this change from the current PR and resolve the issue properly (ideally without the need for `-Wno-return-type` as compile flag).

File: src/cbmc/CMakeLists.txt
Comment (jparsert) [reply src/cbmc/CMakeLists.txt]: Thanks for the suggestion, I changed it to your suggestion.

File: src/cbmc/CMakeLists.txt
Comment (esteffin) [inline src/cbmc/CMakeLists.txt]: Why this line has been tabbed so much?

Similar changes are also everywhere in the file

File: src/cbmc/CMakeLists.txt
Comment (esteffin) [inline src/cbmc/CMakeLists.txt:61]: This is actually one of the 4 lines that should have been modified, but still without the extra padding.

File: src/libcprover-cpp/api.cpp
Comment (esteffin) [inline src/libcprover-cpp/api.cpp]: Why has this file being touched by this PR?
It shouldn't be changed.

File: src/libcprover-cpp/api.h
Comment (esteffin) [inline src/libcprover-cpp/api.h]: Why has this file being touched by this PR?
It shouldn't be changed.

File: src/libcprover-rust/include/c_api.h
Comment (esteffin) [inline src/libcprover-rust/include/c_api.h]: Why has this file being touched by this PR?
It shouldn't be changed.

File: src/libcprover-rust/include/c_errors.h
Comment (esteffin) [inline src/libcprover-rust/include/c_errors.h]: Why has this file being touched by this PR?
It shouldn't be changed.

File: src/libcprover-rust/src/c_api.cc
Comment (esteffin) [inline src/libcprover-rust/src/c_api.cc]: Why has this file being touched by this PR?
It shouldn't be changed.

File: src/libcprover-rust/src/lib.rs
Comment (esteffin) [inline src/libcprover-rust/src/lib.rs]: Why has this file being touched by this PR?
It shouldn't be changed.

File: unit/libcprover-cpp/verification_resultt.cpp
Comment (esteffin) [inline unit/libcprover-cpp/verification_resultt.cpp]: Why has this file being touched by this PR?
It shouldn't be changed.

**Review by esteffin — CHANGES_REQUESTED** [review]
The PR has changes on several unrelated files.
Is this because some other commits have been dragged in when rebasing?
Also there is a huge amount of padding added to `src/cbmc/CMakeLists.txt` that seems due to some automatic linter, but that obfuscate the commit.

Please drop the commits that touch unrelated files and remove the extra padding added.

After this I will be more than happy to approve as the other changes are good.

My suggestion is to totally drop commits `e684a7b04aa22118fea2df3a7d7d721983bcb3aa`, `51d143b9880205a857b4646f9cc3333a3aa43fcf` and `31bccd218dbc63e28727cb0b21e987a6aaf1bf18`. Then to squash `87ce062f83e881c6a1486b5a6b81ad6d33db89b5`, `36a581c37aa88422a1cb41ffd9fe191e57c43979`, `143f97d0515681efdd4de5bb228e5e6332e22525` (after removing the extra padding).
Alternatively this can be closed here and the changes moved to a PR with a cleaner history.

File: src/cbmc/CMakeLists.txt
Comment (jparsert) [reply src/cbmc/CMakeLists.txt]: fixed

File: src/cbmc/CMakeLists.txt
Comment (jparsert) [reply src/cbmc/CMakeLists.txt:61]: fixed

File: src/libcprover-cpp/api.cpp
Comment (jparsert) [reply src/libcprover-cpp/api.cpp]: removed, sorry for cluttering

File: src/libcprover-rust/include/c_api.h
Comment (jparsert) [reply src/libcprover-rust/include/c_api.h]: removed, sorry for cluttering

File: src/libcprover-rust/include/c_errors.h
Comment (jparsert) [reply src/libcprover-rust/include/c_errors.h]: removed, sorry for cluttering

File: src/libcprover-rust/src/c_api.cc
Comment (jparsert) [reply src/libcprover-rust/src/c_api.cc]: removed, sorry for cluttering

File: src/libcprover-rust/src/lib.rs
Comment (jparsert) [reply src/libcprover-rust/src/lib.rs]: removed, sorry for cluttering

File: unit/libcprover-cpp/verification_resultt.cpp
Comment (jparsert) [reply unit/libcprover-cpp/verification_resultt.cpp]: removed, sorry for cluttering

**Comment by jparsert** [comment]
Sorry for adding clutter/commits that should not belong. I cleaned it up and it should now only contain my changes (without padding issues)

**Review by esteffin — APPROVED** [review]
Good for me

**Comment by jparsert** [comment]
Is there anything more I have to do to get pull request accepted?

**Review by thomasspriggs — APPROVED** [review]

**Comment by thomasspriggs** [comment]
> Is there anything more I have to do to get pull request accepted?

The approval from @esteffin should be sufficient to get the cmake change merged. He is a member of the @diffblue/diffblue-opensource group which is codeowner for the `CMakeLists.txt` files.

I think this still needs separate approval for the docs change as neither of your approvals are from documentation code owners.  @NlightNFotis can you approve so we have docs code owner approval?

FYI - Who exactly needs to approve should be defined based on https://github.com/diffblue/cbmc/blob/develop/CODEOWNERS

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
